File: web/src/api/calendarLeaveBy.test.ts
import { describe, expect, it } from "vitest"

import {
  applyLeaveByFillIn,
  mergeCheapCalendarItem,
  mergeCheapCalendarItems,
} from "@/api/calendarLeaveBy"
import type { CalendarItem, CalendarLeaveBy } from "@/api/types"

function item(
  partial: Partial<CalendarItem> & Pick<CalendarItem, "id" | "title">,
): CalendarItem {
  return {
    source: "MANUAL",
    startsAt: "2026-08-15T17:00:00.000Z",
    endsAt: null,
    location: "Rink",
    kidIds: ["k1"],
    feedId: null,
    feedName: null,
    leaveFromPlaceId: "p1",
    leaveFromPlaceName: "Home",
    leaveByAt: null,
    leaveByStatus: "PENDING",
    leaveByReason: null,
    coverages: [],
    uncoveredKidIds: [],
    conflicts: [],
    rsvps: [],
    ...partial,
  }
}

describe("mergeCheapCalendarItem", () => {
  it("uses incoming UNAVAILABLE and clears a stale OK", () => {
    const cached = item({
      id: "e1",
      title: "Practice",
      leaveByStatus: "OK",
      leaveByAt: "2026-08-15T16:00:00.000Z",
      leaveByReason: null,
    })
    const incoming = item({
      id: "e1",
      title: "Practice",
      leaveByStatus: "UNAVAILABLE",
      leaveByReason: "NO_ORIGIN",
      leaveFromPlaceId: null,
      leaveFromPlaceName: null,
    })
    expect(mergeCheapCalendarItem(incoming, cached)).toEqual(incoming)
  })

  it("keeps cached settled leave-by when cheap is PENDING and origin matches", () => {
    const cached = item({
      id: "e1",
      title: "Practice",
      leaveByStatus: "OK",
      leaveByAt: "2026-08-15T16:00:00.000Z",
    })
    const incoming = item({
      id: "e1",
      title: "Practice refreshed",
      leaveByStatus: "PENDING",
      leaveByAt: null,
    })
    const merged = mergeCheapCalendarItem(incoming, cached)
    expect(merged.title).toBe("Practice refreshed")
    expect(merged.leaveByStatus).toBe("OK")
    expect(merged.leaveByAt).toBe("2026-08-15T16:00:00.000Z")
  })

  it("shows PENDING when origin differs from cached OK", () => {
    const cached = item({
      id: "e1",
      title: "Practice",
      leaveFromPlaceId: "p1",
      leaveByStatus: "OK",
      leaveByAt: "2026-08-15T16:00:00.000Z",
    })
    const incoming = item({
      id: "e1",
      title: "Practice",
      leaveFromPlaceId: "p2",
      leaveFromPlaceName: "Dad's",
      leaveByStatus: "PENDING",
      leaveByAt: null,
    })
    expect(mergeCheapCalendarItem(incoming, cached).leaveByStatus).toBe("PENDING")
  })

  it("shows PENDING when there is no cached settled leave-by", () => {
    const incoming = item({ id: "e1", title: "Practice", leaveByStatus: "PENDING" })
    expect(mergeCheapCalendarItem(incoming, undefined).leaveByStatus).toBe("PENDING")
    expect(
      mergeCheapCalendarItem(
        incoming,
        item({ id: "e1", title: "Practice", leaveByStatus: "PENDING" }),
      ).leaveByStatus,
    ).toBe("PENDING")
  })

  it("uses incoming OK from a warm cheap list", () => {
    const incoming = item({
      id: "e1",
      title: "Practice",
      leaveByStatus: "OK",
      leaveByAt: "2026-08-15T16:10:00.000Z",
    })
    const cached = item({
      id: "e1",
      title: "Practice",
      leaveByStatus: "OK",
      leaveByAt: "2026-08-15T16:00:00.000Z",
    })
    expect(mergeCheapCalendarItem(incoming, cached)).toEqual(incoming)
  })
})

describe("mergeCheapCalendarItems", () => {
  it("matches by source and id", () => {
    const cached = [
      item({
        id: "e1",
        title: "Old",
        leaveByStatus: "OK",
        leaveByAt: "2026-08-15T16:00:00.000Z",
      }),
    ]
    const incoming = [
      item({ id: "e1", title: "New", leaveByStatus: "PENDING" }),
      item({ id: "e2", title: "Other", source: "FEED", leaveByStatus: "PENDING" }),
    ]
    const merged = mergeCheapCalendarItems(incoming, cached)
    expect(merged[0]?.leaveByStatus).toBe("OK")
    expect(merged[1]?.leaveByStatus).toBe("PENDING")
  })
})

describe("applyLeaveByFillIn", () => {
  it("overwrites leave-by fields and ignores unknown ids", () => {
    const items = [
      item({ id: "e1", title: "Practice", leaveByStatus: "PENDING" }),
      item({ id: "e2", title: "Game", leaveByStatus: "PENDING" }),
    ]
    const rows: CalendarLeaveBy[] = [
      {
        id: "e1",
        source: "MANUAL",
        leaveFromPlaceId: "p1",
        leaveFromPlaceName: "Home",
        leaveByAt: "2026-08-15T16:20:00.000Z",
        leaveByStatus: "OK",
        leaveByReason: null,
      },
      {
        id: "missing",
        source: "MANUAL",
        leaveByStatus: "OK",
        leaveByAt: "2026-08-15T16:00:00.000Z",
      },
    ]
    const next = applyLeaveByFillIn(items, rows)
    expect(next[0]?.leaveByStatus).toBe("OK")
    expect(next[0]?.leaveByAt).toBe("2026-08-15T16:20:00.000Z")
    expect(next[1]?.leaveByStatus).toBe("PENDING")
  })
})

File: mobile/iosApp/scripts/calendar_leave_by_test.swift
import Foundation

@main
struct CalendarLeaveByTestMain {
    static func main() {
        func expect(_ condition: Bool, _ message: String) {
            if !condition {
                fputs("FAIL: \(message)\n", stderr)
                exit(1)
            }
        }

        func item(
            id: String,
            title: String = "Practice",
            leaveByStatus: String = "PENDING",
            leaveByAt: String? = nil,
            leaveByReason: String? = nil,
            leaveFromPlaceId: String? = "p1",
            source: String = "MANUAL"
        ) -> FamilyCalendarItem {
            FamilyCalendarItem(
                id: id,
                source: source,
                title: title,
                startsAt: "2026-08-15T17:00:00Z",
                endsAt: nil,
                location: "Rink",
                kidIds: ["k1"],
                feedId: nil,
                feedName: nil,
                leaveFromPlaceId: leaveFromPlaceId,
                leaveFromPlaceName: leaveFromPlaceId == "p1" ? "Home" : "Dad's",
                leaveByAt: leaveByAt,
                leaveByStatus: leaveByStatus,
                leaveByReason: leaveByReason
            )
        }

        let cachedOk = item(
            id: "e1",
            leaveByStatus: "OK",
            leaveByAt: "2026-08-15T16:00:00Z"
        )
        let incomingUnavailable = item(
            id: "e1",
            leaveByStatus: "UNAVAILABLE",
            leaveByReason: "NO_ORIGIN",
            leaveFromPlaceId: nil
        )
        expect(
            CalendarLeaveByMerge.mergeCheapCalendarItem(
                incoming: incomingUnavailable,
                cached: cachedOk
            ) == incomingUnavailable,
            "incoming UNAVAILABLE clears stale OK"
        )

        let incomingPending = item(id: "e1", title: "Practice refreshed")
        let kept = CalendarLeaveByMerge.mergeCheapCalendarItem(
            incoming: incomingPending,
            cached: cachedOk
        )
        expect(kept.title == "Practice refreshed", "cheap PENDING keeps refreshed title")
        expect(kept.leaveByStatus == "OK", "cheap PENDING keeps cached OK")
        expect(kept.leaveByAt == "2026-08-15T16:00:00Z", "cheap PENDING keeps cached leaveByAt")

        let originChanged = item(id: "e1", leaveFromPlaceId: "p2")
        expect(
            CalendarLeaveByMerge.mergeCheapCalendarItem(
                incoming: originChanged,
                cached: cachedOk
            ).leaveByStatus == "PENDING",
            "origin change drops stale OK to PENDING"
        )

        expect(
            CalendarLeaveByMerge.mergeCheapCalendarItem(
                incoming: incomingPending,
                cached: nil
            ).leaveByStatus == "PENDING",
            "no cache → PENDING"
        )

        let painted = [
            item(id: "e1"),
            item(id: "e2", title: "Game"),
        ]
        expect(
            LeaveByDisplay.leaveByAgendaLine(
                leaveByStatus: painted[0].leaveByStatus,
                leaveByAt: painted[0].leaveByAt,
                leaveByReason: painted[0].leaveByReason
            ) == LeaveByDisplay.pendingLabel,
            "cheap list paints PENDING before fill-in"
        )
        let filled = CalendarLeaveByMerge.applyLeaveByFillIn(
            items: painted,
            rows: [
                FamilyCalendarLeaveBy(
                    id: "e1",
                    source: "MANUAL",
                    leaveFromPlaceId: "p1",
                    leaveFromPlaceName: "Home",
                    leaveByAt: "2026-08-15T16:20:00Z",
                    leaveByStatus: "OK",
                    leaveByReason: nil
                ),
                FamilyCalendarLeaveBy(
                    id: "missing",
                    source: "MANUAL",
                    leaveFromPlaceId: nil,
                    leaveFromPlaceName: nil,
                    leaveByAt: "2026-08-15T16:00:00Z",
                    leaveByStatus: "OK",
                    leaveByReason: nil
                ),
            ]
        )
        expect(filled[0].leaveByStatus == "OK", "fill-in overwrites matching row")
        expect(filled[0].leaveByAt == "2026-08-15T16:20:00Z", "fill-in leaveByAt")
        expect(filled[1].leaveByStatus == "PENDING", "unknown fill-in ids ignored")
        expect(filled[1].leaveByAt == nil, "omitted row stays PENDING")

        let now = ManualEventDateCodec.date(fromIso: "2026-08-13T18:00:00Z")!
        let loaded = ManualEventDateCodec.defaultCalendarWindow(now: now)
        let near = ManualEventDateCodec.nearTermLeaveByWindow(
            loadedFromIso: loaded.from,
            loadedToIso: loaded.to,
            now: now
        )
        let rest = ManualEventDateCodec.remainderAfterNearTermLeaveByWindow(
            loadedFromIso: loaded.from,
            loadedToIso: loaded.to,
            now: now
        )
        expect(near != nil, "near-term window exists")
        expect(rest != nil, "remainder exists for 30-day window")
        expect(near!.from == loaded.from, "near-term starts at loaded from")
        expect(near!.to == rest!.from, "remainder starts when near-term ends")
        expect(rest!.to == loaded.to, "remainder ends at loaded to")
        let nearStart = ManualEventDateCodec.date(fromIso: near!.from)!
        let nearEnd = ManualEventDateCodec.date(fromIso: near!.to)!
        let nearDays = Calendar.current.dateComponents([.day], from: nearStart, to: nearEnd).day
        expect(nearDays == ManualEventDateCodec.leaveByNearTermDays, "near-term is +2 calendar days")

        let onlyNear = ManualEventDateCodec.advanceCalendarWindow(
            from: ManualEventDateCodec.startOfLocalDay(now),
            days: ManualEventDateCodec.leaveByNearTermDays
        )
        expect(
            ManualEventDateCodec.remainderAfterNearTermLeaveByWindow(
                loadedFromIso: onlyNear.from,
                loadedToIso: onlyNear.to,
                now: now
            ) == nil,
            "remainder is nil when window is only near-term"
        )

        let scriptsDir = URL(fileURLWithPath: #filePath).deletingLastPathComponent()
        let authViewModelURL = scriptsDir
            .deletingLastPathComponent()
            .appendingPathComponent("iosApp/AuthViewModel.swift")
        let authViewModel = try! String(contentsOf: authViewModelURL, encoding: .utf8)
        expect(
            authViewModel.contains("listCalendarLeaveBy"),
            "AuthViewModel fetches leave-by fill-in"
        )
        expect(
            authViewModel.contains("mergeCheapCalendarItems"),
            "cheap list merges with cached leave-by"
        )
        expect(
            authViewModel.contains("calendarItems = CalendarLeaveByMerge.mergeCheapCalendarItems"),
            "Agenda paints cheap/merged items before fill-in"
        )
        let assignRange = authViewModel.range(
            of: "self.calendarItems = CalendarLeaveByMerge.mergeCheapCalendarItems"
        )!
        let fillRange = authViewModel.range(of: "self.fillLeaveByForWindow")!
        expect(
            assignRange.lowerBound < fillRange.lowerBound,
            "paint cheap list before starting leave-by fill-in"
        )
        expect(
            authViewModel.contains("nearTermLeaveByWindow"),
            "near-term slice is requested first"
        )
        expect(
            authViewModel.contains("remainderAfterNearTermLeaveByWindow"),
            "later window fill follows near-term"
        )
        let nearCall = authViewModel.range(of: "nearTermLeaveByWindow")!
        let restCall = authViewModel.range(of: "remainderAfterNearTermLeaveByWindow")!
        expect(
            nearCall.lowerBound < restCall.lowerBound,
            "near-term leave-by request is issued before later-window request"
        )
        expect(
            authViewModel.contains("waitForNearTermFill"),
            "Load more waits for near-term before filling the appended page"
        )
        expect(
            authViewModel.contains("fetchAndApplyLeaveBy(from: page.from, to: page.to"),
            "Load more fills leave-by for the appended page only"
        )
        expect(
            authViewModel.contains("Keep last known leave-by; do not wipe Agenda."),
            "fill-in failure keeps the list intact"
        )
        expect(
            !authViewModel.contains("self.calendarItems = []")
                || authViewModel.contains("onError: { _ in"),
            "leave-by onError must not clear calendarItems"
        )
        let listLeaveByError = authViewModel.range(
            of: "Keep last known leave-by; do not wipe Agenda."
        )!
        let afterError = authViewModel[listLeaveByError.upperBound...]
        let resumeNil = afterError.range(of: "continuation.resume(returning: nil)")
        expect(resumeNil != nil, "fill-in error returns nil rows instead of wiping")

        print("CalendarLeaveBy tests passed")
    }
}
